### PR TITLE
BUG: using .ix with a multi-index indexer

### LINKED ISSUE #11372: DataFrame.ix[idx, :] = value sets wrong values when idx is a MultiIndex and DataFrame.columns is also a MultiIndex
This code is broken in `0.17.0` but not in `0.15.2`:

```python
import pandas as pd
import numpy as np

np.random.seed(1)

from itertools import product

from pandas.util.testing import assert_frame_equal

pd.show_versions()

idx = pd.MultiIndex.from_tuples(
    list(
        product(['A', 'B', 'C'], 
                pd.date_range('2015-01-01', '2015-04-01', freq='MS'))
    )
)

sub = pd.MultiIndex.from_tuples(
    [('A', pd.Timestamp('2015-01-01')), ('A', pd.Timestamp('2015-02-01'))]
)
# if cols = ['foo', 'bar', 'baz', 'quux'], there is no error. 
cols = pd.MultiIndex.from_tuples(
    list(
        product(['foo', 'bar'], 
                pd.date_range('2015-01-01', '2015-02-01', freq='MS'))
    )
)

test = pd.DataFrame(np.random.random((12, 4)), index=idx, columns=cols)
vals = pd.DataFrame(np.random.random((2, 4)), index=sub, columns=cols)
test.ix[sub, :] = vals

print test.ix[sub, :]
print vals

assert_frame_equal(test.ix[sub, :], vals)
```

### 0.17.0

```python
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 26 Stepping 5, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0.17.0
nose: 1.3.7
pip: 7.1.0
setuptools: 18.0.1
Cython: 0.22
numpy: 1.10.1
scipy: 0.16.0
statsmodels: 0.6.1
IPython: 3.2.1
sphinx: 1.3.1
patsy: 0.4.0
dateutil: 2.4.1
pytz: 2015.4
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.4.4
matplotlib: 1.4.3
openpyxl: None
xlrd: 0.9.4
xlwt: None
xlsxwriter: 0.7.3
lxml: None
bs4: 4.3.2
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: 1.0.7
pymysql: None
psycopg2: None
                    foo                   bar
             2015-01-01 2015-02-01 2015-01-01 2015-02-01
A 2015-01-01   0.287775   0.130029   0.019367   0.678836
  2015-02-01   0.287775   0.130029   0.019367   0.678836
                    foo                   bar
             2015-01-01 2015-02-01 2015-01-01 2015-02-01
A 2015-01-01   0.287775   0.130029   0.019367   0.678836
  2015-02-01   0.211628   0.265547   0.491573   0.053363
Traceback (most recent call last):
  File "c:\dev\code\sandbox\multiindex.py", line 41, in <module>
    assert_frame_equal(test.ix[sub, :], vals)
  File "c:\python\envs\pd017\lib\site-packages\pandas\util\testing.py", line 1028, in assert_frame_equal
    obj='DataFrame.iloc[:, {0}]'.format(i))
  File "c:\python\envs\pd017\lib\site-packages\pandas\util\testing.py", line 925, in assert_series_equal
    check_less_precise, obj='{0}'.format(obj))
  File "pandas\src\testing.pyx", line 58, in pandas._testing.assert_almost_equal (pandas\src\testing.c:3809)
  File "pandas\src\testing.pyx", line 147, in pandas._testing.assert_almost_equal (pandas\src\testing.c:2685)
  File "c:\python\envs\pd017\lib\site-packages\pandas\util\testing.py", line 798, in raise_assert_detail
    raise AssertionError(msg)
AssertionError: DataFrame.iloc[:, 0] are different

DataFrame.iloc[:, 0] values are different (50.0 %)
[left]:  [0.287775338586, 0.287775338586]
[right]: [0.287775338586, 0.211628116]
```

### 0.15.2
```python

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 26 Stepping 5, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en_GB

pandas: 0.15.2
nose: 1.3.7
Cython: 0.22
numpy: 1.9.2
scipy: 0.15.1
statsmodels: None
IPython: 3.2.1
sphinx: 1.3.1
patsy: 0.3.0
dateutil: 2.4.1
pytz: 2015.4
bottleneck: 1.0.0
tables: 3.2.0
numexpr: 2.4.3
matplotlib: 1.4.3
openpyxl: 1.8.5
xlrd: 0.9.4
xlwt: 0.7.5
xlsxwriter: 0.7.3
lxml: 3.4.4
bs4: 4.3.2
html5lib: 0.999
httplib2: None
apiclient: None
rpy2: None
sqlalchemy: 1.0.7
pymysql: None
psycopg2: None
                    foo                   bar           
             2015-01-01 2015-02-01 2015-01-01 2015-02-01
A 2015-01-01   0.287775   0.130029   0.019367   0.678836
  2015-02-01   0.211628   0.265547   0.491573   0.053363
                    foo                   bar           
             2015-01-01 2015-02-01 2015-01-01 2015-02-01
A 2015-01-01   0.287775   0.130029   0.019367   0.678836
  2015-02-01   0.211628   0.265547   0.491573   0.053363
```

### PR DESCRIPTION
**Pull request by rekcahpassyla**
closes #11372 

Add an optional argument to _NDFrameIndexer to indicate
if the indexer is from a MultiIndex

### CODE REVIEW

File: pandas/tests/test_indexing.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from numpy import nan
 from numpy.random import randn
 import numpy as np
+from numpy.testing import assert_allclose
```
Comment (jreback) [inline pandas/tests/test_indexing.py]: we don't use this

File: pandas/tests/test_indexing.py [hunk 2]
```diff
@@ -788,6 +789,50 @@ def test_loc_setitem_multiindex(self):
...
+    def test_ix_setitem_multiindex(self):
+        # GH5206
+        df = pd.DataFrame(np.random.random((5, 5)), columns='A,B,C,D,E'.split(','))
+        df['F'] = 99
+        row_selection = df['A']<0.5
+
+        col_selection = ['B','C']
+
+        df.ix[row_selection, col_selection] = df['F']
+        assert_allclose(df.ix[row_selection, col_selection], 99)
```
Comment (jreback) [inline pandas/tests/test_indexing.py]: use ``tm.assert_almost_equal``

File: pandas/tests/test_indexing.py [hunk 3]
```diff
@@ -788,6 +789,50 @@ def test_loc_setitem_multiindex(self):
         result = df.loc[(t,n),'X']
         self.assertEqual(result,3)
 
+    def test_ix_setitem_multiindex(self):
+        # GH5206
+        df = pd.DataFrame(np.random.random((5, 5)), columns='A,B,C,D,E'.split(','))
+        df['F'] = 99
+        row_selection = df['A']<0.5
+
```
Comment (jreback) [inline pandas/tests/test_indexing.py]: don't use extra whitespace

File: pandas/tests/test_indexing.py [hunk 4]
```diff
@@ -788,6 +789,50 @@ def test_loc_setitem_multiindex(self):
         result = df.loc[(t,n),'X']
         self.assertEqual(result,3)
 
+    def test_ix_setitem_multiindex(self):
+        # GH5206
+        df = pd.DataFrame(np.random.random((5, 5)), columns='A,B,C,D,E'.split(','))
```
Comment (jreback) [inline pandas/tests/test_indexing.py]: test for ``.ix`` and ``.loc`` (you can make a check function to simplify)

File: pandas/tests/test_indexing.py [hunk 5]
```diff
@@ -788,6 +789,50 @@ def test_loc_setitem_multiindex(self):
...
+        row_selection = df['A']<0.5
+
+        col_selection = ['B','C']
+
+        df.ix[row_selection, col_selection] = df['F']
+        assert_allclose(df.ix[row_selection, col_selection], 99)
+
+        # GH11372
+        idx = pd.MultiIndex.from_tuples(list(
+            itertools.product(
```
Comment (jreback) [inline pandas/tests/test_indexing.py]: use ``pd.MultiIndex.from_product``

File: pandas/tests/test_indexing.py [hunk 6]
```diff
@@ -788,6 +789,50 @@ def test_loc_setitem_multiindex(self):
...
+            [('A', pd.Timestamp('2015-01-01')), ('A', pd.Timestamp('2015-02-01'))]
+        )
+
+        subcols = pd.MultiIndex.from_tuples(
+            [('foo', pd.Timestamp('2016-01-01')), ('foo', pd.Timestamp('2016-02-01'))]
+        )
+
+        vals = pd.DataFrame(np.random.random((2, 2)), index=subidx, columns=subcols)
+        test.ix[subidx, subcols] = vals
+        assert_frame_equal(test.ix[subidx, subcols], vals)
```
Comment (jreback) [inline pandas/tests/test_indexing.py]: add a test for identity (e.g. test.ix[test.index,test.columns] = test

File: pandas/tests/test_indexing.py (hunk 4 above)
Comment (rekcahpassyla) [inline pandas/tests/test_indexing.py]: I'm also going to change this test to not use `np.random` I think, as I note my row selection is `<0.5` below, and thus the test inputs will be different between runs, which seems like a bad smell.

**Comment by rekcahpassyla** [comment]
Updated with comments from review. I have consolidated the tests `test_ix_setitem_multiindex` and `test_loc_setitem_multiindex` into `test_setitem_multiindex`

**Comment by jreback** [comment]
pls add a whatsnew note (bug fixes), looks good otherwise.

**Comment by rekcahpassyla** [comment]
Thanks - rebased on master, added whatsnew and pushed.

File: pandas/core/indexing.py [hunk 7]
```diff
@@ -443,11 +443,14 @@ def can_do_equal_len():
                 # we have an equal len Frame
                 if isinstance(value, ABCDataFrame) and value.ndim > 1:
                     sub_indexer = list(indexer)
+                    indexer_multiindex = isinstance(labels, MultiIndex)
```
Comment (jreback) [inline pandas/core/indexing.py]: how about rename this to ``multiiindex_indexer`` (for a bit of consistency)

File: pandas/core/indexing.py [hunk 8]
```diff
@@ -516,8 +519,10 @@ def can_do_equal_len():
             self.obj._data = self.obj._data.setitem(indexer=indexer, value=value)
             self.obj._maybe_update_cacher(clear=True)
 
-    def _align_series(self, indexer, ser):
+    def _align_series(self, indexer, ser, indexer_multiindex=False):
         # indexer to assign Series can be tuple, slice, scalar
+        # indexer_multiindex should be set to True if the indexer came from
+        # a multi index. Otherwise, the items will be broadcast unnecessarily
         if isinstance(indexer, (slice, np.ndarray, list, Index)):
             indexer = tuple([indexer])
```
Comment (jreback) [inline pandas/core/indexing.py]: same here.

can you update the doc-string as well (e.g. with a ``Parameters/Returns`` section)

**Comment by rekcahpassyla** [comment]
Renamed to `multiindex_indexer` and added docstring for `_align_series`

**Comment by jreback** [comment]
@rekcahpassyla thanks! 

that was a nice deep dive. lots of other indexing issues if you'd like :smile:

**Comment by rekcahpassyla** [comment]
Welcome! In the short term, a bit pressed for time but we are testing our entire library against `0.17.0` and if anything else comes out of that, I'll be happy to have a crack at those too.

**Pull request merged**